gcloud/storage/batch.py

```diff
@@ -0,0 +1,208 @@
+
+        for method, uri, headers, body in deferred:
+            sub = MIMEApplicationHTTP(method, uri, headers, body)
+            multi.attach(sub)
+
+        # Flatten payload
+        if six.PY3:             # pragma: NO COVER  Python3
+            buf = io.StringIO()
+        else:                   # pragma: NO COVER  Python2
+            buf = io.BytesIO()
+        generator = Generator(buf, False, 0)
+        generator.flatten(multi)
+        payload = buf.getvalue()
+
+        # Strip off redundant header text
+        _, body = payload.split('\n\n', 1)
+        headers = dict(multi._headers)
+
+        url = self._connection.build_api_url('/batch')
+
+        _req = self._connection._make_request
+        response, content = _req('POST', url, data=payload, headers=headers)
+        self._responses = list(_crack_mime_response(response, content))
+        return self._responses
+
+    def __enter__(self):
+        _BATCHES.push(self)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        try:
+            if exc_type is None:
+                self.finish()
+            else:
+                self._http.reset()
+        finally:
+            _BATCHES.pop()
+
+
+def _crack_mime_response(response, content):
+    """Convert response, content -> [(status, reason, payload)].
+    """
+    parser = Parser()
+    faux = ('Content-Type: %s\nMIME-Version: 1.0\n\n%s' %
+            (response['Content-Type'], content))
+
+    message = parser.parsestr(faux)
+
+    if not isinstance(message._payload, list):
+        raise ValueError('Bad response:  not multi-part')
+
+    for sub in message._payload:
+        status_line, rest = sub._payload.split('\n', 1)
+        _, status, reason = status_line.split(' ', 2)
+        message = parser.parsestr(rest)
+        payload = message._payload
+        ctype = message['Content-Type']
+        if ctype and ctype.startswith('application/json'):
+            payload = json.loads(payload)
+        yield status, reason, payload
+
+
+class NoContent(object):
+    """Emulate an HTTP '204 No Content' response."""
+    status = 204
+
+
+class _FauxHTTP(object):
+    """Emulate ``connection.http``, but store requests.
+
+    Only allow up to ``_MAX_BATCH_SIZE`` requests to be bathed.
+    """
+    _MAX_BATCH_SIZE = 1000
+
+    def __init__(self, connection):
+        self._connection = connection
+        self._requests = []
+        self._orig_http, connection.http = connection.http, self
+
+    def request(self, method, uri, headers, body):
+        """Emulate / proxy underlying HTTP request.
+
+        - Pass 'GET' requests through.
+
+        - Defer others for later processing
+        """
+        if method == 'GET':
+            _req = self._orig_http.request
+            return _req(method=method, uri=uri, headers=headers, body=body)
+
+        if len(self._requests) >= self._MAX_BATCH_SIZE:
+            raise ValueError("Too many deferred requests (max %d)" %
```

Comment:
This is a reason why I'd rather not monkey patch `connection.http`. If someone caught this and moved on, `reset()` would never be called.